### PR TITLE
[TS migration] Migrate tests in react-native-onyx

### PR DESCRIPTION
**Pull request by blazejkustra**
### Details

This PR aims to migrate all Onyx tests to Typescript.

### Related Issues

[GH_LINK](https://github.com/Expensify/App/issues/40355)

### Automated Tests

N/A

### Manual Tests

N/A

### Author Checklist

- [x] I linked the correct issue in the `### Related Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

File: tests/unit/subscribeToPropertiesTest.tsx [hunk 1]
```diff
@@ -138,7 +139,7 @@ describe('Only the specific property changes when using withOnyx() and ', () =>
                     Onyx.mergeCollection(ONYX_KEYS.COLLECTION.TEST_KEY, {
                         [`${ONYX_KEYS.COLLECTION.TEST_KEY}1`]: {a: 'one', b: 'two'},
                         [`${ONYX_KEYS.COLLECTION.TEST_KEY}2`]: {c: 'three', d: 'four'},
-                    }),
+                    } as Collection<string, unknown, unknown>),
```
Comment (VickyStash) [inline tests/unit/subscribeToPropertiesTest.tsx:142]: `Collection<string, unknown, unknown>` is used a lot, maybe it makes sense to create a named type for it

File: tests/unit/withOnyxTest.tsx [hunk 2]
```diff
@@ -265,15 +268,16 @@ describe('withOnyxTest', () => {
...
             text: {
                 key: 'test',
                 initWithStoredValues: false,
             },
         })(ViewWithText);
         TestComponentWithOnyx.defaultProps = {
             text: INITIAL_VALUE,
-        };
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        } as any;
```
Comment (VickyStash) [inline tests/unit/withOnyxTest.tsx:280]:
```suggestion
        } as ViewWithTextProps;
```

File: tests/components/ViewWithText.tsx [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
+import {View, Text} from 'react-native';
+
+type ViewWithTextProps = {
+    text: string | null;
+    // eslint-disable-next-line react/no-unused-prop-types
+    collectionID?: string;
```
Comment (VickyStash) [inline tests/components/ViewWithText.tsx:7]: Let's add a clarification why it's needed

File: tests/unit/onyxTest.ts [hunk 4]
```diff
@@ -733,8 +739,9 @@ describe('Onyx', () => {
                         },
                     ],
                 },
-            ]);
-        } catch (error) {
+            ] as unknown as OnyxUpdate[]);
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        } catch (error: any) {
             expect(error.message).toEqual('Invalid value provided in Onyx multiSet. Onyx multiSet value must be of type object.');
         }
```
Comment (VickyStash) [inline tests/unit/onyxTest.ts:746]:
```suggestion
        } catch (error) {
            if (error instanceof Error) {
               expect(error.message).toEqual('Invalid value provided in Onyx multiSet. Onyx multiSet value must be of type object.');
            }
        }
```

File: tests/unit/onyxTest.ts [hunk 5]
```diff
@@ -663,15 +667,17 @@ describe('Onyx', () => {
...
 
             // When we pass it to Onyx.update
+            // @ts-expect-error This is an invalid call to Onyx.update
             Onyx.update(data);
-        } catch (error) {
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        } catch (error: any) {
             // Then we should expect the error message below
             expect(error.message).toEqual('Invalid boolean key provided in Onyx update. Onyx key must be of type string.');
         }
```
Comment (VickyStash) [inline tests/unit/onyxTest.ts:676]:
```suggestion
        } catch (error) {
            if (error instanceof Error) {
               // Then we should expect the error message below
               expect(error.message).toEqual('Invalid boolean key provided in Onyx update. Onyx key must be of type string.');
            }
        }
```

File: tests/unit/withOnyxTest.tsx [hunk 6]
```diff
@@ -742,15 +756,15 @@ describe('withOnyxTest', () => {
...
+        const TestComponentWithOnyx = withOnyx<ViewWithTextProps, {text: unknown}>({
             // Note: the prop passed to the wrapped component is called "text",
             // which is different than the key selected by the selector "hello"
             text: {
                 key: ONYX_KEYS.TEST_KEY,
                 selector,
             },
         })(ViewWithText);
-        let sourceData = {hello: 'world', goodnight: 'moon'};
+        let sourceData: unknown = {hello: 'world', goodnight: 'moon'};
```
Comment (VickyStash) [inline tests/unit/withOnyxTest.tsx:767]:
```suggestion
        let sourceData: OnyxValue<typeof ONYX_KEYS.TEST_KEY> = {hello: 'world', goodnight: 'moon'};
```

File: tests/utils/waitForPromisesToResolve.ts [hunk 7]
```diff
@@ -1,4 +1,4 @@
-export default () =>
+export default (): Promise<unknown> =>
```
Comment (VickyStash) [inline tests/utils/waitForPromisesToResolve.ts:1]: Why not `Promise<void>`?

File: tests/unit/fastMergeTest.ts [hunk 8]
```diff
@@ -82,13 +85,15 @@ describe('fastMerge', () => {
     });
 
     it('should replace an object with an array', () => {
-        const result = utils.fastMerge(testObject, [1, 2, 3]);
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        const result = utils.fastMerge(testObject, [1, 2, 3] as any);
```
[the remark this answers is not in the document]
Comment (blazejkustra) [reply tests/unit/fastMergeTest.ts:89]: I tried updating the types when I was migrating Onyx.js, unfortunately I gave up 😅

File: tests/unit/onyxTest.ts (hunk 4 above)
Comment (blazejkustra) [reply tests/unit/onyxTest.ts:746]: I changed it, but I think we might need to throw an error if it's not instanceOf Error:
```ts
   if (error instanceof Error) {
                expect(error.message).toEqual('Invalid value provided in Onyx multiSet. Onyx multiSet value must be of type object.');
            } else {
                throw error;
            }
```

File: tests/unit/withOnyxTest.tsx (hunk 6 above)
Comment (blazejkustra) [reply tests/unit/withOnyxTest.tsx:766]: It's actually the same as `OnyxValue<typeof ONYX_KEYS.TEST_KEY>` === `unknown`. I agree it looks better

**Comment by blazejkustra** [comment]
@VickyStash Thank you for your attention to details! Adjusted 😄

File: tests/unit/onyxTest.ts [hunk 9]
```diff
@@ -645,15 +648,17 @@ describe('Onyx', () => {
...
             {onyxMethod: 'murge', key: ONYX_KEYS.OTHER_TEST, value: {test2: 'test2'}},
         ];
 
         try {
             // When we pass it to Onyx.update
+            // @ts-expect-error This is an invalid call to Onyx.update
             Onyx.update(data);
-        } catch (error) {
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        } catch (error: any) {
```
Comment (VickyStash) [inline tests/unit/onyxTest.ts:661]: Let's get rid of `any` here as well using `if (error instanceof Error) {` check

**Review by VickyStash — APPROVED** [review]
LGTM

**Review by tgolen — APPROVED** [review]

**Pull request merged**